tiatoolbox/annotation/storage.py

```diff
@@ -0,0 +1,1709 @@
+            where=where,
+        )
+        if isinstance(where, Callable):
+            return [
+                Annotation(self._unpack_geometry(blob, cx, cy), properties)
+                for blob, properties, cx, cy in cur.fetchall()
+                if where(json.loads(properties))
+            ]
+        return [
+            Annotation(
+                self._unpack_geometry(blob, cx, cy),
+                json.loads(properties),
+            )
+            for blob, properties, cx, cy in cur.fetchall()
+        ]
+
+    def __len__(self) -> int:
+        cur = self.con.cursor()
+        cur.execute("SELECT COUNT(*) FROM annotations")
+        (count,) = cur.fetchone()
+        return count
+
+    def __contains__(self, key: str) -> bool:
+        cur = self.con.cursor()
+        cur.execute("SELECT EXISTS(SELECT 1 FROM annotations WHERE [key] = ?)", (key,))
+        return cur.fetchone()[0] == 1
+
+    def __getitem__(self, key: str) -> Annotation:
+        cur = self.con.cursor()
+        cur.execute(
+            """
+            SELECT boundary, properties, cx, cy
+              FROM annotations
+             WHERE [key] = :key
+            """,
+            {"key": key},
+        )
+        row = cur.fetchone()
+        if row is None:
+            raise KeyError(key)
+        boundary, properties, cx, cy = row
+        properties = json.loads(properties or "{}")
+        geometry = self._unpack_geometry(boundary, cx, cy)
+        return Annotation(geometry, properties)
+
+    def keys(self) -> Iterable[int]:
+        yield from self
+
+    def __iter__(self) -> Iterable[int]:
+        cur = self.con.cursor()
+        cur.execute(
+            """
+            SELECT [key]
+              FROM annotations
+            """
+        )
+        while True:
+            row = cur.fetchone()
+            if row is None:
+                break
+            key = row[0]
+            yield key
+
+    def values(self) -> Iterable[Tuple[int, Annotation]]:
+        for _, value in self.items():
+            yield value
+
+    def items(self) -> Iterable[Tuple[int, Annotation]]:
+        cur = self.con.cursor()
+        cur.execute(
+            """
+            SELECT [key], cx, cy, boundary, properties
+              FROM annotations
+            """
+        )
+        while True:
+            row = cur.fetchone()
+            if row is None:
+                break
+            key, cx, cy, boundary, properties = row
+            if boundary is not None:
+                geometry = self._unpack_geometry(boundary, cx, cy)
+            else:
+                geometry = Point(cx, cy)
+            properties = json.loads(properties)
+            yield key, Annotation(geometry, properties)
+
+    def patch_many(
+        self,
+        keys: Iterable[int],
+        geometries: Optional[Iterable[Geometry]] = None,
+        properties_iter: Optional[Iterable[Properties]] = None,
+    ) -> None:
+        # Validate inputs
```

Comment:
docstring missing.

Reply:
This function should inherit the docstring of the parent classe. It can be copied out if desierd but it will be done automatically by sphinx.